### ISSUE TITLE
Many `vec256<float>` operations  returns incorrect result on aarch64 if compiled by gcc-7.5.0

### ISSUE THREAD

**Issue by malfet** [pytorch/pytorch#47098]
## To Reproduce

1. Install Ubuntu-18.04 + gcc-7.5.0
1. Build Pytorch
1. Run `python3 -c "import torch;print(torch.hypot(torch.tensor([1.]), torch.tensor([1.])))"`

The result would be a surprising `tensor([1.])`.

## Expected behavior

Hypotenuse of unit triangle should be close to square root of 2, i.e.:
```
$ python3 -c "import torch;print(torch.hypot(torch.tensor([1.]), torch.tensor([1.])))"
tensor([1.4142])
```
## Environment

```
$ python3 -mtorch.utils.collect_env
Collecting environment information...
PyTorch version: 1.8.0a0+e3b55a8
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (aarch64)
GCC version: (Ubuntu/Linaro 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.6 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.3
[pip3] torch==1.8.0a0
[conda] Could not collect

```

**Comment by malfet** [comment]
And problem disappears if compiled by gcc-8.4.0 or newer

**Comment by malfet** [comment]
Updated the title to reflect that issue is not limited to `hypot`, but affects lots of operators accelerated with `neon` intrinsics. 
For example 37 tests from https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/vec256_test_all_types.cpp, namely:
```
[  FAILED  ] Rounding/0.Round, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Rounding/0.Ceil, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Rounding/0.Floor, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Trigonometric/0.Sin, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Trigonometric/0.Cos, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Trigonometric/0.Tan, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Hyperbolic/0.Tanh, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Hyperbolic/0.Sinh, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Hyperbolic/0.Cosh, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] InverseTrigonometric/0.Asin, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] InverseTrigonometric/0.ACos, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] InverseTrigonometric/0.ATan, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Logarithm/0.Log, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Logarithm/0.Log10, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] LogarithmReals/0.Log2, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] LogarithmReals/0.Log1p, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Exponents/0.Exp, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Exponents/0.Expm1, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] ErrorFunctions/0.Erf, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] ErrorFunctions/0.Erfc, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] InverseTrigonometricReal/0.ATan2, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Pow/0.Pow, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] RealTests/0.Hypot, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Interleave/0.Interleave, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Interleave/0.DeInterleave, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Arithmetics/0.Plus, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Arithmetics/0.Minus, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Arithmetics/0.Multiplication, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Bitwise/0.BitAnd, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Bitwise/0.BitOr, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Bitwise/0.BitXor, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Comparison/0.Equal, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Comparison/0.NotEqual, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Comparison/0.Greater, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Comparison/0.Less, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Comparison/0.GreaterEqual, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] Comparison/0.LessEqual, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
[  FAILED  ] BitwiseFloatsAdditional/0.ZeroMask, where TypeParam = at::vec256::(anonymous namespace)::Vec256<float>
```

**Comment by gchanan** [comment]
@glaringlee do the vec256 tests cover this?

**Comment by gchanan** [comment]
also @malfet -- do we ship binaries with these problems?

**Comment by malfet** [comment]
@gchanan We didn't ship aarch64 binaries for 1.7, but are about to do so soon.

There are another set of accuracy problems with NEON, will submit a separate issue about that.

And failures above are from vec256 tests.

**Comment by glaringlee** [comment]
@malfet @gchanan 
I don't think we were running vec256 test cases (in `vec256_test_all_types.h/cpp`) under ARM before. What we tested including gnu, clang, and windows, but not including mobile. `

We have `vec256_test.cpp` which does very basic test when Neon vec was added. I am trying to merge vec256_test.cpp into vec256_test_all_types.cpp (for non mobile build at the moment) since they did lots of overlapping tests, I was wondering whether vec256_test_all_types.cpp can run under ARM earlier today, it seems above info gave me the answer :). 

The vec256_test_all_types.h/cpp did tweaks for different platforms, sometimes, we have to lower precision, or manually prevent compiler optimization (eg, prevent `fmadd` optimization) to achieve the testing goals without problem. Not sure whether we have internal env to build pytorch under ARM, if so, we can dig this and fix the vec256 test.
